subversion - fix stack trace when getting repository information

samdoran opened the pull request:
##### SUMMARY
Fixes #36498.

When `LC_ALL` is not set, the output language of `svn` commands will differ based on locale. There is a lot of history of trying to fix this. See the following pull requests:

https://github.com/ansible/ansible-modules-core/pull/3257
https://github.com/ansible/ansible-modules-core/pull/4358

This patch attempts to fix both issues by setting `LC_ALL` to a UTF-8 locale. Setting `LC_ALL` to `C` reintroduces
this bug: https://github.com/ansible/ansible-modules-core/issues/4178.

I'm sure there are some problems I am not seeing with setting this to `en_US.UTF-8`, but that is the only way I could find to fix this bug without reintroducing the bug in https://github.com/ansible/ansible-modules-core/issues/4178.


##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/modules/subversion.py`

samdoran committed:
subversion - set LC_ALL for accurate command output parsing

When LC_ALL is not set, the output language of commands will differ based on locale. There
is a lot of history of trying to fix this. See the following pull requests:

https://github.com/ansible/ansible-modules-core/pull/4358
https://github.com/ansible/ansible-modules-core/pull/4358

This patch attempts to fix this my setting LC_ALL to a UTF-8 locale. Setting LC_ALL to C reintroduces
this bug https://github.com/ansible/ansible-modules-core/issues/4178.

I'm sure there are some problems I am not seeing with setting this to en_US.UTF-8, but that is
the only way I could find to fix this bug without reintriducing the bug mentioned above.

bcoca commented on lib/ansible/modules/subversion.py:
```diff
@@ -295,7 +295,7 @@ def main():
     # We screenscrape a huge amount of svn commands so use C locale anytime we
     # call run_command()
-    module.run_command_environ_update = dict(LANG='C', LC_MESSAGES='C')
+    module.run_command_environ_update = dict(LANG='C', LC_MESSAGES='C', LC_ALL='en_US.UTF-8')
```
C/POSIX is always available en_US is not guaranteed

samdoran replied:
Ok. And the reason I needed use a UTF-8 locale is because the paths are not being converted to bytes, as we discussed elsewhere. I'll work on that on Monday.

sivel replied:
We could potentially attempt `C.UTF-8`, but that may not exist either.  It looks like fedora/rhel may not have always had it: https://bugzilla.redhat.com/show_bug.cgi?id=902094

Some research may need to be done.

bcoca replied:
that is a bit newer, so in the future it will become the default/failover, but plenty of 'old' installs out there so we cannot rely on it yet.

samdoran replied:
After converting the repo path to bytes, I am still getting the same error from `subversion` when the repo path contains non-ASCII characters if `LC_ALL=C`. The error below from subversion appears when the modules runs `svn info /tmp/tésting/xml-crimson`.

```
svn: E000022: Can't convert string from native encoding to 'UTF-8':
svn: E000022: /tmp/testing/xml-crimson
```

From reading the [documentation on how subversion handles localization](http://svnbook.red-bean.com/en/1.2/svn.advanced.l10n.html) I believe the issue is that subversion does not like that the current system locale does not support UTF-8.

> Errors like this typically occur when the Subversion client has received a UTF-8 string from the repository, but not all of the characters in that string can be represented using the encoding of the current locale. 

That makes me think that we won't be able to normalize the messages _and_ use paths that contain non-ASCII characters. 😞

bcoca replied:
weee! 

So I would try C.UTF-8 and fallback to C if that fails, so unicode paths will work on newer machines and we can explain rest and even give option to 'force en_US.UTF-8' if user knows its available.

bcoca replied:
or people just need to roll out their i18n version of subversion module to deal with messages in their native lang+encoding

samdoran replied:
I looked at adjusting the regexp that searches for revisions and managed to get it working for all the locales on my system. But that seems like a pretty inaccurate fix. And I would need to do it for other regexp searches to make it complete...

samdoran replied:
I talked to @mattclay about how best to test this and that ended up being a whole other big topic. Currently we don't have any tests running Ansible under a non-English locale. So the test would need to install the locale and configure it to make sure it's available (as @bcoca mentioned earlier).

I looked into using `C.UTF-8` and it's also pretty complicated, particularly for BSD and macOS. [PEP-538](https://www.python.org/dev/peps/pep-0538/) and [PEP-540](https://www.python.org/dev/peps/pep-0540/) provide some insight into how Python handled this.

samdoran committed:
Rather than setting locale, just check for matches before trying to get groups

This is a pragmatic solution to avoid the stack trace since setting the locale correctly
to ensure message parsing is accurate is problematic.

samdoran committed:
Improve regexps for finding revision and URL

samdoran committed:
Update changelog

samdoran commented:
I decided to go with a pragmatic fix. It's not perfect, but better than how it works currently.

samdoran commented:
/rebuild

samdoran commented:
/rebuild

samdoran commented:
/azp run

samdoran commented:
/azp run
